File: typescript/packages/core/src/core/google/date_glob.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { GLOB_CHARS } from '../../utils/glob_walk.ts'

function day(year: number, month: number, date: number): string {
  const mm = String(month).padStart(2, '0')
  const dd = String(date).padStart(2, '0')
  return `${String(year)}-${mm}-${dd}`
}

function isValidDate(year: number, month: number, date: number): boolean {
  if (month < 1 || month > 12 || date < 1) return false
  const d = new Date(Date.UTC(year, month - 1, date))
  return d.getUTCFullYear() === year && d.getUTCMonth() === month - 1 && d.getUTCDate() === date
}

function parseFixedInt(s: string | undefined, expectedLength: number): number | null {
  if (s?.length !== expectedLength || !/^\d+$/.test(s)) return null
  return Number.parseInt(s, 10)
}

/**
 * Translate a date-prefixed glob into a half-open range of floating dates.
 *
 * Kept separate from `globToModifiedRange` because a caller bucketing in a
 * named time zone has to build its own bounds from the dates; UTC instants
 * would silently shift the window by the zone's offset.
 */
export function globToDateRange(pattern: string | null | undefined): [string, string] | null {
  if (!pattern) return null
  let metaIndex = -1
  for (const ch of GLOB_CHARS) {
    const idx = pattern.indexOf(ch)
    if (idx !== -1 && (metaIndex === -1 || idx < metaIndex)) metaIndex = idx
  }
  if (metaIndex === -1) return null
  const prefix = pattern.slice(0, metaIndex).replace(/[_-]+$/, '')
  const parts = prefix.split('-')
  if (parts.length === 1) {
    const year = parseFixedInt(parts[0], 4)
    if (year === null) return null
    return [day(year, 1, 1), day(year + 1, 1, 1)]
  }
  if (parts.length === 2) {
    const year = parseFixedInt(parts[0], 4)
    const month = parseFixedInt(parts[1], 2)
    if (year === null || month === null) return null
    if (!isValidDate(year, month, 1)) return null
    if (month === 12) return [day(year, month, 1), day(year + 1, 1, 1)]
    return [day(year, month, 1), day(year, month + 1, 1)]
  }
  if (parts.length === 3) {
    const year = parseFixedInt(parts[0], 4)
    const month = parseFixedInt(parts[1], 2)
    const date = parseFixedInt(parts[2], 2)
    if (year === null || month === null || date === null) return null
    if (!isValidDate(year, month, date)) return null
    const next = new Date(Date.UTC(year, month - 1, date) + 86400000)
    return [
      day(year, month, date),
      day(next.getUTCFullYear(), next.getUTCMonth() + 1, next.getUTCDate()),
    ]
  }
  return null
}

/** Translate a date-prefixed glob into an RFC3339 modifiedTime range. */
export function globToModifiedRange(pattern: string | null | undefined): [string, string] | null {
  const span = globToDateRange(pattern)
  if (span === null) return null
  return [`${span[0]}T00:00:00Z`, `${span[1]}T00:00:00Z`]
}

File: python/mirage/core/google/date_glob.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from datetime import date, timedelta

from mirage.utils.glob_walk import GLOB_CHARS


def _iso(d: date) -> str:
    return d.strftime("%Y-%m-%dT00:00:00Z")


def glob_to_modified_range(pattern: str | None) -> tuple[str, str] | None:
    if not pattern:
        return None
    meta_index = -1
    for ch in GLOB_CHARS:
        idx = pattern.find(ch)
        if idx != -1 and (meta_index == -1 or idx < meta_index):
            meta_index = idx
    if meta_index == -1:
        return None
    prefix = pattern[:meta_index]
    prefix = prefix.rstrip("_-")
    parts = prefix.split("-")
    try:
        if len(parts) == 1 and len(parts[0]) == 4:
            year = int(parts[0])
            start = date(year, 1, 1)
            end = date(year + 1, 1, 1)
        elif len(parts) == 2 and len(parts[0]) == 4 and len(parts[1]) == 2:
            year = int(parts[0])
            month = int(parts[1])
            start = date(year, month, 1)
            if month == 12:
                end = date(year + 1, 1, 1)
            else:
                end = date(year, month + 1, 1)
        elif (len(parts) == 3 and len(parts[0]) == 4 and len(parts[1]) == 2
              and len(parts[2]) == 2):
            year = int(parts[0])
            month = int(parts[1])
            day = int(parts[2])
            start = date(year, month, day)
            end = start + timedelta(days=1)
        else:
            return None
    except ValueError:
        return None
    return _iso(start), _iso(end)
